File: 2022/seminars/01-dslib/ping-pong/basic.py
from dslib import Context, Message, Node


class PingClient(Node):
    def __init__(self, node_id: str, server_id: str):
        self._id = node_id
        self._server_id = server_id

    def on_local_message(self, msg: Message, ctx: Context):
        if msg.type == 'PING':
            ctx.send(msg, self._server_id)

    def on_message(self, msg: Message, sender: str, ctx: Context):
        # process messages from server
        if msg.type == 'PONG':
            ctx.send_local(msg)

    def on_timer(self, timer_name: str, ctx: Context):
        # process fired timers here
        pass


class PingServer(Node):
    def __init__(self, node_id: str):
        self._id = node_id

    def on_local_message(self, msg: Message, ctx: Context):
        # not used
        pass

    def on_message(self, msg: Message, sender: str, ctx: Context):
        # process messages from client
        if msg.type == 'PING':
            pong = Message('PONG', {'value': msg['value']})
            ctx.send(pong, sender)

    def on_timer(self, timer_name: str, ctx: Context):
        # process fired timers here
        pass

File: week-01/anysystem-intro/ping-pong/tests/src/solution/basic.rs
use anysystem::{Context, Message, Process};

#[derive(Clone)]
pub struct Client {
    server: String,
}

impl Client {
    pub fn new(server: &str) -> Self {
        Self {
            server: server.to_string(),
        }
    }
}

impl Process for Client {
    fn on_message(&mut self, msg: Message, _from: String, ctx: &mut Context) -> Result<(), String> {
        if msg.tip == "PONG" {
            ctx.send_local(msg);
        }
        Ok(())
    }

    fn on_local_message(&mut self, msg: Message, ctx: &mut Context) -> Result<(), String> {
        if msg.tip == "PING" {
            ctx.send(msg, self.server.clone());
        }
        Ok(())
    }

    fn on_timer(&mut self, _timer: String, _ctx: &mut Context) -> Result<(), String> {
        Ok(())
    }
}

#[derive(Clone)]
pub struct Server {}

impl Server {
    pub fn new() -> Self {
        Self {}
    }
}

impl Process for Server {
    fn on_message(&mut self, msg: Message, from: String, ctx: &mut Context) -> Result<(), String> {
        if msg.tip == "PING" {
            let resp = Message::new("PONG".to_string(), msg.data);
            ctx.send(resp, from);
        }
        Ok(())
    }

    fn on_local_message(&mut self, _msg: Message, _ctx: &mut Context) -> Result<(), String> {
        Ok(())
    }

    fn on_timer(&mut self, _timer: String, _ctx: &mut Context) -> Result<(), String> {
        Ok(())
    }
}
